core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -676,52 +676,72 @@ private RexNode simplifyCoalesce(RexCall call) {
 
   private RexNode simplifyCase(RexCall call, RexUnknownAs unknownAs) {
     List<CaseBranch> inputBranches =
-        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList(call.getOperands()));
+        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList<>(call.getOperands()));
 
     // run simplification on all operands
-    RexSimplify branchSimplifier = this;
-    RelDataType branchType = call.getType();
+    RexSimplify condSimplifier = this.withPredicates(RelOptPredicateList.EMPTY);
+    RexSimplify valueSimplifier = this;
+    RelDataType caseType = call.getType();
 
+    boolean merged = false;
```

Comment:
In my point of view `conditionNeedsSimplify` would make more sense

Reply:
Makes sense, I changed it.